moveit_planners/trajopt/src/problem_description.cpp

```diff
@@ -0,0 +1,567 @@
+    first_step = last_step;
+    last_step = tmp;
+    ROS_WARN("Last time step for JointPosTerm comes before first step. Reversing them.");
+  }
+  if (last_step == -1)  // last_step not set
+    last_step = first_step;
+
+  // Check if parameters are the correct size.
+  checkParameterSize(coeffs, n_dof, "JointPoseTermInfo coeffs", true);
+  checkParameterSize(targets, n_dof, "JointPoseTermInfo targets", true);
+  checkParameterSize(upper_tols, n_dof, "JointPoseTermInfo upper_tols", true);
+  checkParameterSize(lower_tols, n_dof, "JointPoseTermInfo lower_tols", true);
+
+  // Check if tolerances are all zeros
+  bool is_upper_zeros =
+      std::all_of(upper_tols.begin(), upper_tols.end(), [](double i) { return util::doubleEquals(i, 0.); });
+  bool is_lower_zeros =
+      std::all_of(lower_tols.begin(), lower_tols.end(), [](double i) { return util::doubleEquals(i, 0.); });
+
+  // Get vars associated with joints
+  trajopt::VarArray vars = prob.GetVars();
+  trajopt::VarArray joint_vars = vars.block(0, 0, vars.rows(), static_cast<int>(n_dof));
+  if (prob.GetHasTime())
+    ROS_INFO("JointPoseTermInfo does not differ based on setting of TT_USE_TIME");
+
+  if (term_type & TT_COST)
+  {
+    // If the tolerances are 0, an equality cost is set. Otherwise it's a hinged "inequality" cost
+    if (is_upper_zeros && is_lower_zeros)
+    {
+      prob.addCost(sco::CostPtr(new trajopt::JointPosEqCost(joint_vars, util::toVectorXd(coeffs),
+                                                            util::toVectorXd(targets), first_step, last_step)));
+      prob.getCosts().back()->setName(name);
+    }
+    else
+    {
+      prob.addCost(sco::CostPtr(new trajopt::JointPosIneqCost(joint_vars, util::toVectorXd(coeffs),
+                                                              util::toVectorXd(targets), util::toVectorXd(upper_tols),
+                                                              util::toVectorXd(lower_tols), first_step, last_step)));
+      prob.getCosts().back()->setName(name);
+    }
+  }
+  else if (term_type & TT_CNT)
+  {
+    // If the tolerances are 0, an equality cnt is set. Otherwise it's an inequality constraint
+    if (is_upper_zeros && is_lower_zeros)
+    {
+      prob.addConstraint(sco::ConstraintPtr(new trajopt::JointPosEqConstraint(
+          joint_vars, util::toVectorXd(coeffs), util::toVectorXd(targets), first_step, last_step)));
+      prob.getConstraints().back()->setName(name);
+    }
+    else
+    {
+      prob.addConstraint(sco::ConstraintPtr(new trajopt::JointPosIneqConstraint(
+          joint_vars, util::toVectorXd(coeffs), util::toVectorXd(targets), util::toVectorXd(upper_tols),
+          util::toVectorXd(lower_tols), first_step, last_step)));
+      prob.getConstraints().back()->setName(name);
+    }
+  }
+  else
+  {
+    ROS_WARN("JointPosTermInfo does not have a valid term_type defined. No cost/constraint applied");
+  }
+}
+
+void JointVelTermInfo::hatch(TrajOptProblem& prob)
```

Comment:
can we use a better name than hatch in all of these?

Reply:
`addObjectiveTerms` instead? @BryceStevenWilley ?

Reply:
love it